src/actions/mintNFT.ts

```diff
@@ -1,3 +1,152 @@
-export const mintNFT = async () => {
-  throw new Error("It's not implemented");
+import { MintLayout, TOKEN_PROGRAM_ID, Token } from '@solana/spl-token';
+import { Keypair, PublicKey } from '@solana/web3.js';
+import { Connection } from './../Connection';
+import { MintTo, CreateAssociatedTokenAccount, CreateMint, Transaction } from './../programs';
+import {
+  CreateMasterEdition,
+  CreateMetadata,
+  Creator,
+  MasterEdition,
+  Metadata,
+  MetadataDataData,
+} from './../programs/metadata';
+import { Wallet } from 'src/wallet';
+import { ASSOCIATED_TOKEN_PROGRAM_ID } from '@solana/spl-token';
+import { sendTransaction } from './transactions';
+import { lookup } from './../utils/metadata';
+import BN from 'bn.js';
+
+interface MintNFTParams {
+  connection: Connection;
+  wallet: Wallet;
+  uri: string;
+  maxSupply: number;
+}
+
+interface MintNFTResponse {
+  txId: string;
+  mint: PublicKey;
+  metadata: PublicKey;
+  edition: PublicKey;
+}
+
+export const mintNFT = async ({
+  connection,
+  wallet,
+  uri,
+  maxSupply,
+}: MintNFTParams): Promise<MintNFTResponse> => {
+  const mint = Keypair.generate();
+
+  const metadataPDA = await Metadata.getPDA(mint.publicKey);
+  const editionPDA = await MasterEdition.getPDA(mint.publicKey);
+
+  const mintRent = await connection.getMinimumBalanceForRentExemption(MintLayout.span);
+
+  const {
+    name,
+    symbol,
+    seller_fee_basis_points,
+    properties: { creators },
+  } = await lookup(uri);
+
+  const creatorsData = creators.reduce((memo, { address, share }) => {
```

Comment:
To prevent type casting

```suggestion
  const creatorsData = creators.reduce<Creator>((memo, { address, share }) => {
```